src/parser.cpp

```diff
@@ -376,54 +365,21 @@ namespace ada::parser {
    */
   std::optional<uint64_t> parse_ipv4_number(std::string_view input) {
     // If input is the empty string, then return failure.
-    if (input.empty()) {
-      return std::nullopt;
+    if (input.empty()) { return std::nullopt; }
+    // using std::strtoll is potentially dangerous if the string is not null terminated.
+    uint64_t result{};
+    std::from_chars_result r;
+    if((input.length() >= 2) && ((input[0] == '0') & (checkers::to_lower(input[1]) == 'x'))) {
```

Comment:
Why not `&&`?

```suggestion
    if((input.length() >= 2) && ((input[0] == '0') && (checkers::to_lower(input[1]) == 'x'))) {
```

Reply:
Let us look at how these functions get compiled...

```C++
bool z(std::string_view input) {
    return((input.length() >= 2) && ((input[0] == '0') && (to_lower(input[1]) == 'x')));
}


bool z2(std::string_view input) {
    return((input.length() >= 2) && ((input[0] == '0') & (to_lower(input[1]) == 'x')));
}
```

With '&&' (the first one), you have

```asm
        xor     eax, eax
        cmp     rdi, 1
        jbe     .L1
        cmp     BYTE PTR [rsi], 48
        je      .L6
.L1:
        ret
.L6:
        movzx   eax, BYTE PTR [rsi+1]
        or      eax, 32
        cmp     al, 120
        sete    al
        ret
```

With the second one, you have...

```asm
        xor     eax, eax
        cmp     rdi, 1
        jbe     .L7
        movzx   eax, BYTE PTR [rsi+1]
        or      eax, 32
        cmp     al, 120
        sete    al
        cmp     BYTE PTR [rsi], 48
        sete    dl
        and     eax, edx
```

With the second one, you save a branch. Saving a branch is a loss if the branch is often taken. If you almost always have something that starts with '0', then it is better to branch (have '&&'), if not, you may prefer not to branch.

So we should look at the actual data and see which one is expected to be the best decision.

Branch predictors are very good on modern hardware... but that's also a big problem because it gives the illusion that branchy code is faster on small benchmarks (like ours) because we repeat the same tests in a closed loop. So I tend to default on branchless code even if it seems a tiny bit slower on small benchmarks, because I expect that in the real world, when the processor can't learn the branches, I will do much better.

Reply:
But actually, we can do much better and do just one check... let me see...

Reply:
There you go...
```C++
bool has_hex_prefix(std::string_view input) {
    uint32_t value = 1;
    bool is_little_endian = (static_cast<uint8_t>(value) == 1);
    uint16_t word0x{};
    std::memcpy(&word0x, "0x", 2); // we would use bit_cast in C++20 and the function could be constexpr.
    uint16_t two_first_bytes{};
    std::memcpy(&two_first_bytes, input.data(),2);
    if(is_little_endian) { two_first_bytes |= 0x2000; } else { two_first_bytes |= 0x020; }
    return two_first_bytes == word0x;
  }
```

compiles down to...

```asm
        movzx   eax, word ptr [rsi]
        and     eax, 57343
        cmp     eax, 22576
        sete    al
        ret
```

which is basically unbeatable I think!!!

Reply:
Nice!!!! Good work. Inspiring! @lemire